### PR TITLE
MythicMob spawns event

### PR DESCRIPTION
**Pull request by Ph4i1ur3**
adds the spawn event for MythicMobs

### CODE REVIEW

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsDeathEvent.java [hunk 1]
```diff
@@ -32,6 +32,8 @@
     //
     // @Regex ^on mythicmob [^\s]+ (dies|death|killed)( by [^\s]+)?( in ((notable (cuboid|ellipsoid))|([^\s]+)))?$
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsDeathEvent.java:33]: fix the other meta too (the event lines and the regex still have inline in area)

Comment (Ph4i1ur3) [reply src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsDeathEvent.java:33]: Oof, will do

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    public static MythicMobsSpawnEvent instance;
+    public MythicMobSpawnEvent event;
+    public MythicMobsMobTag mob;
+    public EntityTag entity;
+    public LocationTag location;
+
+    @Override
+    public boolean couldMatch(ScriptPath path) {
+        return (path.eventLower.startsWith("mythicmob") && (path.eventArgLowerAt(2).equals("spawns")));
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:58]: stray parens

Also, use the standard couldMatch format used through Denizen, not this weird inline thing.

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        String mob = path.eventArgLowerAt(1);
+
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:64]: stray blank lines

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        if (!mob.equals("mob")
+                && !mob.equals(CoreUtilities.toLowerCase(this.mob.getMobType().getInternalName()))) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:66]: redundant newline

Also use the BukkitScriptEvent generic match calls

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        else if (name.equals("entity")) {
+            return entity;
+        }
+        else if (name.equals("location")) {
+            return location;
+        }
+        else if (name.equals("from_spawner")) {
+            return new ElementTag(event.isFromMythicSpawner());
+        }
+        else if (name.equals("spawner_location")) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:96]: This should probably validate isFromMythicSpawner

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            AbstractLocation loc = event.getMythicSpawner().getLocation();
+            return new LocationTag(loc.getX(), loc.getY(), loc.getZ(),loc.getWorld().getName());
+        }
+        return super.getContext(name);
+    }
+
+    @EventHandler
+    public void onMythicMobSpawns(MythicMobSpawnEvent event) {
+        mob = new MythicMobsMobTag(event.getMob());
+        entity = new EntityTag(event.getEntity());
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:106]: `entity` can be just grabbed in the context method not here (other fields probably can too?)

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    // mythicmob <mob> spawns
+    //
+    // @Regex ^on mythicmob [^\s]+ spawns$
+    //
+    // @Switch in:<area> to only process the event if it occurred within a specified area.
+    //
+    // @Triggers when a MythicMob spawns.
+    //
+    // @Context
+    // <context.mob> Returns the MythicMobTag that is spawning.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:29]: `MythicMobsMobTag` (or just `mythic mob`)

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    // <context.entity> Returns the EntityTag for the MythicMob.
+    // <context.location> Returns a LocationTag of where the MythicMob will spawn.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:31]: **the** location, not 'a'

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    // <context.from_spawner> Returns true if the mob was from a spawner.
+    // <context.spawner_location> Returns a LocationTag of the spawner that spawned the mob.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:33]: `, if any.`

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    public static MythicMobsSpawnEvent instance;
+    public MythicMobSpawnEvent event;
+    public MythicMobsMobTag mythicmob;
+    public LocationTag location;
+
+    @Override
+    public boolean couldMatch(ScriptPath path) {
+        return (path.eventLower.startsWith("mythicmob")
+                && (path.eventArgLowerAt(2).equals("spawns")));
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:53]: Again, this is misformatted - check the formatting of any `couldMatch` method in Denizen.

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        String mob = path.eventArgLowerAt(1);
+        if (!mob.equals("mob")
+                && runGenericCheck(mob, mythicmob.getMobType().getInternalName())) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:60]: weird newline

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+            return new EntityTag(event.getEntity());
+        }
+        else if (name.equals("location")) {
+            return location;
+        }
+        else if (name.equals("from_spawner")) {
+            return new ElementTag(event.isFromMythicSpawner());
+        }
+        else if (name.equals("spawner_location")) {
+            if (event.isFromMythicSpawner()) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:89]: use `&&` to combine two `if` statements properly...

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    public static MythicMobsSpawnEvent instance;
+    public MythicMobSpawnEvent event;
+    public MythicMobsMobTag mythicmob;
+    public LocationTag location;
+
+    @Override
+    public boolean couldMatch(ScriptPath path) {
+        String cmd = path.eventArgLowerAt(1);
+        return (path.eventLower.startsWith("mythicmob")
+                && (cmd.equals("spawns")));
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:54]: For the third time with this comment:
Please reference any existing couldMatch method in Denizen

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        else if (name.equals("entity")) {
+            return new EntityTag(event.getEntity());
+        }
+        else if (name.equals("location")) {
+            return location;
+        }
+        else if (name.equals("from_spawner")) {
+            return new ElementTag(event.isFromMythicSpawner());
+        }
+        else if ((name.equals("spawner_location")) && (event.isFromMythicSpawner())) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:88]: weird stray parens

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    public boolean couldMatch(ScriptPath path) {
+        String cmd = path.eventArgLowerAt(2);
+        if (!path.eventLower.startsWith("mythicmob")) {
+            return false;
+        }
+        if (cmd.equals("spawns")) {
+            return false;
+        };
+        return true;
+    };
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:60]: where did all the semicolons here come from

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    public MythicMobsMobTag mythicmob;
+    public LocationTag location;
+
+    @Override
+    public boolean couldMatch(ScriptPath path) {
+        String cmd = path.eventArgLowerAt(2);
+        if (!path.eventLower.startsWith("mythicmob")) {
+            return false;
+        }
+        if (cmd.equals("spawns")) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:56]: This should probably have a `!` on it

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    public static MythicMobsSpawnEvent instance;
+    public MythicMobSpawnEvent event;
+    public MythicMobsMobTag mythicmob;
+    public LocationTag location;
+
+    @Override
+    public boolean couldMatch(ScriptPath path) {
+        String cmd = path.eventArgLowerAt(2);
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:52]: This variable doesn't need to exist

File: src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+        if (cmd.equals("spawns")) {
+            return false;
+        };
+        return true;
+    };
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        String mob = path.eventArgLowerAt(1);
+        if (!mob.equals("mob") && runGenericCheck(mob, mythicmob.getMobType().getInternalName())) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/depenizen/bukkit/events/mythicmobs/MythicMobsSpawnEvent.java:65]: This also should have a `!`

**Pull request merged**